File: Manipulator3D_Ruby/src/robot/robot_arm.rb
# frozen_string_literal: true

require_relative "../raylib_bootstrap"
require_relative "../util/math3d"

module Robot
  class LinkParams
    attr_accessor :length_m, :mass_kg, :inertia_cm, :inertia_joint

    def initialize(length_m: 2.5, mass_kg: 1.0)
      @length_m = length_m.to_f
      @mass_kg = mass_kg.to_f
      @inertia_cm = 0.0
      @inertia_joint = 0.0
      recompute_inertia
    end

    def recompute_inertia
      @inertia_cm = (1.0 / 12.0) * @mass_kg * @length_m * @length_m
      @inertia_joint = (1.0 / 3.0) * @mass_kg * @length_m * @length_m
    end
  end

  class JointAngles
    attr_accessor :q0_yaw, :q1_pitch, :q2_pitch

    def initialize(q0_yaw: 0.0, q1_pitch: 0.0, q2_pitch: 0.0)
      @q0_yaw = q0_yaw.to_f
      @q1_pitch = q1_pitch.to_f
      @q2_pitch = q2_pitch.to_f
    end

    def dup
      JointAngles.new(q0_yaw: @q0_yaw, q1_pitch: @q1_pitch, q2_pitch: @q2_pitch)
    end
  end

  class FKResult
    attr_accessor :base, :joint1, :joint2, :ee

    def initialize
      @base = Vector3.create(0.0, 0.0, 0.0)
      @joint1 = Vector3.create(0.0, 0.0, 0.0)
      @joint2 = Vector3.create(0.0, 0.0, 0.0)
      @ee = Vector3.create(0.0, 0.0, 0.0)
    end
  end

  class IKResult
    attr_accessor :reachable, :q, :message

    def initialize(reachable: false, q: JointAngles.new, message: "")
      @reachable = reachable
      @q = q
      @message = message
    end
  end

  class RobotArm
    def initialize(link1, link2)
      @link1 = link1
      @link2 = link2
      @link1.recompute_inertia
      @link2.recompute_inertia
    end

    attr_reader :link1, :link2

    def l1 = @link1.length_m
    def l2 = @link2.length_m

    def max_reach = l1 + l2
    def min_reach = (l1 - l2).abs

    def solve_ik(target, elbow_up: false)
      if target.z < 0.0
        return IKResult.new(reachable: false, message: "Invalid target: z must be >= 0")
      end

      x = target.x.to_f
      y = target.y.to_f
      z = target.z.to_f

      d = Math.sqrt(x * x + y * y + z * z)
      rmin = min_reach
      rmax = max_reach

      if d < rmin - 1e-9 || d > rmax + 1e-9
        msg = "Target radius |p|=#{d} is outside [#{rmin}, #{rmax}]"
        return IKResult.new(reachable: false, message: msg)
      end

      q0 = (x.abs > 1e-12 || y.abs > 1e-12) ? Math.atan2(y, x) : 0.0

      r = Math.sqrt(x * x + y * y)
      l1v = l1
      l2v = l2

      c2 = (r * r + z * z - l1v * l1v - l2v * l2v) / (2.0 * l1v * l2v)
      c2 = [[c2, -1.0].max, 1.0].min

      q2 = Math.acos(c2)
      q2 = -q2 unless elbow_up

      s2 = Math.sin(q2)
      k1 = l1v + l2v * Math.cos(q2)
      k2 = l2v * s2

      q1 = Math.atan2(z, r) - Math.atan2(k2, k1)

      IKResult.new(
        reachable: true,
        q: JointAngles.new(q0_yaw: q0, q1_pitch: q1, q2_pitch: q2),
        message: "OK"
      )
    end

    def forward_kinematics(q)
      fk = FKResult.new
      fk.base = Vector3.create(0.0, 0.0, 0.0)
      fk.joint1 = fk.base

      cy = Math.cos(q.q0_yaw)
      sy = Math.sin(q.q0_yaw)

      u = Vector3.create(cy, sy, 0.0)
      k = Vector3.create(0.0, 0.0, 1.0)

      p1 = Vector3.create(
        u.x * (l1 * Math.cos(q.q1_pitch)) + k.x * (l1 * Math.sin(q.q1_pitch)),
        u.y * (l1 * Math.cos(q.q1_pitch)) + k.y * (l1 * Math.sin(q.q1_pitch)),
        u.z * (l1 * Math.cos(q.q1_pitch)) + k.z * (l1 * Math.sin(q.q1_pitch))
      )

      a = q.q1_pitch + q.q2_pitch
      p2 = Vector3.create(
        p1.x + u.x * (l2 * Math.cos(a)) + k.x * (l2 * Math.sin(a)),
        p1.y + u.y * (l2 * Math.cos(a)) + k.y * (l2 * Math.sin(a)),
        p1.z + u.z * (l2 * Math.cos(a)) + k.z * (l2 * Math.sin(a))
      )

      fk.joint2 = p1
      fk.ee = p2
      fk
    end
  end
end

File: Manipulator3D_Julia/src/robot/RobotArm.jl
module Robot

export Vec3, v3, norm3,
       LinkParams, JointAngles, FKResult, IKResult,
       RobotArm, min_reach, max_reach,
       solve_ik, forward_kinematics

using Printf

# -----------------------------
# Basic 3D vector (standalone)
# -----------------------------

struct Vec3
    x::Float32
    y::Float32
    z::Float32
end

v3(x, y, z) = Vec3(Float32(x), Float32(y), Float32(z))

Base.:+(a::Vec3, b::Vec3) = Vec3(a.x + b.x, a.y + b.y, a.z + b.z)
Base.:-(a::Vec3, b::Vec3) = Vec3(a.x - b.x, a.y - b.y, a.z - b.z)
Base.:*(s::Real, v::Vec3) = Vec3(Float32(s) * v.x, Float32(s) * v.y, Float32(s) * v.z)
Base.:*(v::Vec3, s::Real) = s * v

function norm3(v::Vec3)::Float32
    return Float32(sqrt(v.x*v.x + v.y*v.y + v.z*v.z))
end

# -----------------------------
# Link parameters
# -----------------------------

struct LinkParams
    length_m::Float32
    mass_kg::Float32
    inertia_cm::Float32
    inertia_joint::Float32
end

"""
Create link parameters for a uniform rod model.

- inertia about center of mass (axis ⟂ to rod): I_cm = (1/12) m L^2
- inertia about joint at one end:              I_joint = (1/3)  m L^2
"""
function LinkParams(; length_m=2.5, mass_kg=1.0)
    L = Float32(length_m)
    m = Float32(mass_kg)
    icm = Float32((1/12) * m * L * L)
    ij  = Float32((1/3)  * m * L * L)
    return LinkParams(L, m, icm, ij)
end

# -----------------------------
# Joint / kinematics structs
# -----------------------------

struct JointAngles
    q0_yaw::Float32   # rad
    q1_pitch::Float32 # rad
    q2_pitch::Float32 # rad
end

JointAngles() = JointAngles(0.0f0, 0.0f0, 0.0f0)

struct FKResult
    base::Vec3
    joint1::Vec3
    joint2::Vec3
    ee::Vec3
end

struct IKResult
    reachable::Bool
    q::JointAngles
    message::String
end

# -----------------------------
# Robot arm model (no recursion)
# -----------------------------

struct RobotArm
    link1::LinkParams
    link2::LinkParams
end

# Reach limits
min_reach(arm::RobotArm) = abs(arm.link1.length_m - arm.link2.length_m)
max_reach(arm::RobotArm) = arm.link1.length_m + arm.link2.length_m

# -----------------------------
# Analytic inverse kinematics
# -----------------------------

"""
Analytic IK for a yaw + planar 2-link arm.

Rules enforced:
- target.z >= 0
- ||p|| within [min_reach, max_reach]
"""
function solve_ik(arm::RobotArm, target::Vec3; elbow_up::Bool=false)::IKResult
    # z constraint
    if target.z < 0.0f0
        return IKResult(false, JointAngles(), "Invalid target: z must be >= 0")
    end

    x = Float64(target.x)
    y = Float64(target.y)
    z = Float64(target.z)

    d = sqrt(x*x + y*y + z*z)
    rmin = Float64(min_reach(arm))
    rmax = Float64(max_reach(arm))

    if d < rmin - 1e-9 || d > rmax + 1e-9
        msg = @sprintf("Target radius |p|=%.6f is outside [%.6f, %.6f]", d, rmin, rmax)
        return IKResult(false, JointAngles(), msg)
    end

    # Base yaw
    q0 = (abs(x) > 1e-12 || abs(y) > 1e-12) ? atan(y, x) : 0.0

    # Reduce to planar (r, z)
    r = sqrt(x*x + y*y)
    L1 = Float64(arm.link1.length_m)
    L2 = Float64(arm.link2.length_m)

    # Elbow from law of cosines
    c2 = (r*r + z*z - L1*L1 - L2*L2) / (2.0 * L1 * L2)
    c2 = clamp(c2, -1.0, 1.0)

    q2 = acos(c2)
    if !elbow_up
        q2 = -q2
    end

    s2 = sin(q2)
    k1 = L1 + L2*cos(q2)
    k2 = L2*s2

    q1 = atan(z, r) - atan(k2, k1)

    q = JointAngles(Float32(q0), Float32(q1), Float32(q2))
    return IKResult(true, q, "OK")
end

# -----------------------------
# Forward kinematics
# -----------------------------

function forward_kinematics(arm::RobotArm, q::JointAngles)::FKResult
    base = v3(0,0,0)
    joint1 = base

    L1 = arm.link1.length_m
    L2 = arm.link2.length_m

    cy = Float32(cos(q.q0_yaw))
    sy = Float32(sin(q.q0_yaw))

    # radial direction in XY and +Z axis
    u = v3(cy, sy, 0)
    k = v3(0, 0, 1)

    # elbow
    p1 = (L1 * cos(q.q1_pitch)) * u + (L1 * sin(q.q1_pitch)) * k

    # end effector
    a = q.q1_pitch + q.q2_pitch
    p2 = p1 + (L2 * cos(a)) * u + (L2 * sin(a)) * k

    return FKResult(base, joint1, p1, p2)
end

end # module Robot
